Add dlna_dms integration to support DLNA Digital Media Servers

chishm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new integration, `dlna_dms`, to incorporate DLNA Digital Media Servers (DMS) as Media Sources.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/21624

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chishm committed:
New: dlna_dms; Support for DLNA Digital Media Servers

balloob commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,232 @@
+            LOGGER.debug("user_input: %s", user_input)
+            source_id = user_input[CONF_ENTITY_ID]
+            # NOTE: source_id is not checked for uniqueness because we don't
+            # have access from here to the other config entries to check (no
```
There should be `self.hass`

chishm replied:
I've dropped configuration of source IDs completely now, so this block of code is gone.

balloob commented on homeassistant/components/dlna_dms/dms.py:
```diff
@@ -0,0 +1,715 @@
+                return False
+            self.entities[entity.unique_id] = entity
+            self.sources[entity.source_id] = entity
+        await entity.async_added_to_hass()
```
Ah, this is not a Home Assistant entity object.

Can you rename the class. I was confused because the called functions have same names as normal entities.

chishm replied:
I can see how that's confusing. I've renamed it now.

For background, this was originally based on the idea of [`MediaSourceEntity`](https://github.com/home-assistant/architecture/discussions/675), which was rejected because media sources haven't got states to report or services to call, unlike other HA entities.

balloob commented on homeassistant/components/dlna_dms/dms.py:
```diff
@@ -0,0 +1,715 @@
+        return self._device is not None and self._device.profile_device.available
+
+    @property
+    def unique_id(self) -> str:
+        """Get the unique ID for this entity, which is the USN."""
```
Whyn ot call this `device_id` ?

chishm replied:
Dropped in favour of using the `config_entry.unique_id` in `DlnaDmsData`, which makes much more sense in context.

balloob commented on homeassistant/components/dlna_dms/dms.py:
```diff
@@ -0,0 +1,715 @@
+
+        # Based on DmrDevice.media_image_url from async_upnp_client.
+        if album_art_uri := getattr(item, "album_art_uri", None):
+            return self._device.get_absolute_url(album_art_uri)
```
Not all absolute URLs can be resolved. For example, if you're outside the network. 

Plex has solved this by adding it's own web view to fall back to when the request is not from the internal network: https://github.com/home-assistant/core/pull/66702/files#diff-4bc5cfdbcf43f8ad35f24f9c6ae347141b2bd4ee85a26bc372b411422e52337bR79-R86

This could be left for an extra PR.

chishm replied:
Is it worth doing the browser thumbnails proxying in `media_player`, the same way it proxies the currently playing media's thumbnail?

balloob replied:
Media player doesn't know if a URL can be locally resolved or not.

chishm replied:
Ah, that makes sense. I'll work on this in a future PR.

balloob commented on homeassistant/components/dlna_dms/dms.py:
```diff
@@ -0,0 +1,715 @@
+        return f"{self.source_id}/{action}{object_id}"
+
+
+class Action(str, enum.Enum):
```
`StrEnum` can be imported from `homeassistant.backports.enum`.

balloob commented on homeassistant/components/dlna_dms/media_source.py:
```diff
@@ -0,0 +1,134 @@
+    async def async_resolve_media(self, item: MediaSourceItem) -> DidlPlayMedia:
+        """Resolve a media item to a playable item."""
+        dms_data = self._get_domain_data()
+        if not dms_data:
+            raise Unresolvable("dlna_dms has not been configured")
```
This is not possible. The DLNA DMS media source is only loaded if the setup succeeded.

balloob commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,232 @@
+            # NOTE: source_id is not checked for uniqueness because we don't
+            # have access from here to the other config entries to check (no
+            # hass object reference).
+            options[CONF_ENTITY_ID] = source_id
```
- Don't call this entity ID.
- Why do we allow configuring this? Let's just slugify the name and call it a day.

balloob commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,232 @@
+        LOGGER.debug("async_step_user: user_input: %s", user_input)
+
+        if user_input is not None and (host := user_input.get(CONF_HOST)):
+            # User has chosen a device, ask for confirmation
```
You're not asking for confirmation (and I don't think you should)
```suggestion
            # User has chosen a device
```

balloob reviewed (changes requested):
Looks great! Left a couple of comments.

chishm committed:
Fix issues from code review

* Rename DmsEntity -> DmsDeviceSource; it's not a HA Entity object and the
  old name is confusing
* Remove configuration of source ID; derive it directly from the config
  entry title, which can be configured
* No need to check for missing domain data from within media source
  functions; it must be configured for them to be called

balloob commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,171 @@
+        }
+
+        data_schema = vol.Schema(
+            {vol.Optional(CONF_HOST): vol.In(self._discoveries.keys())}
```
You can pass a dictionary here, and it will use the values as labels. So you could map hostname -> UPNP_FRIENDLY_NAME

chishm replied:
Thank you for the hint, I didn't know that was a possibility.

balloob commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,171 @@
+            return self.async_abort(reason="no_devices_found")
+
+        self._discoveries = {
+            discovery.upnp.get(ssdp.ATTR_UPNP_FRIENDLY_NAME)
+            or cast(str, urlparse(discovery.ssdp_location).hostname): discovery
```
This feels a bit weird, why not key it by USN ?

chishm replied:
I didn't know I could use a mapping in the schema, so I tried to make the displayed keys as human-friendly as possible. Also, UPnP devices may not supply a friendly name, so there was a fallback to the hostname. All fixed now.

balloob commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,171 @@
+        """Handle a flow initialized by SSDP discovery."""
+        LOGGER.debug("async_step_ssdp: discovery_info %s", pformat(discovery_info))
+
+        await self._async_set_info_from_discovery(discovery_info)
```
Should you move this after where we know the device can be used ?

chishm replied:
Probably not; `_async_set_info_from_discovery` does some early parsing of the `discovery_info` and does the checks for unique IDs. I've renamed the method to reflect its actual role and added a comment, so it's hopefully less confusing now.

balloob commented on homeassistant/components/dlna_dms/const.py:
```diff
@@ -0,0 +1,118 @@
+MEDIA_TYPE_MAP: Mapping[str, str] = {
+    "object": _mp_const.MEDIA_TYPE_URL,
+    "object.item": _mp_const.MEDIA_TYPE_URL,
+    "object.item.imageItem": _mp_const.MEDIA_TYPE_IMAGE,
```
For media source the media content type should be the actual mime type.

chishm committed:
More code review fixes

* Better listing of discovered devices for user config flow
* Use MIME type or UPnP class for media_content_type

chishm commented:
Thank you for the code reviews. I think I've fixed all the problems found.

balloob reviewed (approved):
🎉

chishm commented:
@balloob Here's the promised discussion about media metadata: https://github.com/home-assistant/architecture/discussions/739

MartinHjelmare commented on homeassistant/components/dlna_dms/dms.py:
```diff
@@ -0,0 +1,726 @@
+class DmsDeviceSource:
+    """DMS Device wrapper, providing media files as a media_source."""
+
+    hass: HomeAssistant
```
We only need to type attributes here if it differs from the type annotated in the init method for the attributes.

MartinHjelmare commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class ConnectError(IntegrationError):
```
We should not inherit from `IntegrationError` here. That exception is used for a special group of exceptions by our helpers. Please inherit `HomeAssistantError` instead.

MartinHjelmare replied:
`ConnectError` doesn't seem used.

MartinHjelmare commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+    def _create_entry(self) -> FlowResult:
+        """Create a config entry, assuming all required information is now known."""
+        LOGGER.debug(
+            "_async_create_entry: location: %s, USN: %s", self._location, self._usn
```
There's no method named `_async_create_entry`.

MartinHjelmare commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+            self._location = discovery_info.ssdp_location
+
+        self._usn = discovery_info.ssdp_usn
+        await self.async_set_unique_id(self._usn)
```
When setting the unique id in the user flow we need to pass `raise_on_progress=False`. We always want user flows to take precedence over discovery flows.

MartinHjelmare commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+
+    async def _async_get_discoveries(self) -> list[ssdp.SsdpServiceInfo]:
+        """Get list of unconfigured DLNA devices discovered by SSDP."""
+        LOGGER.debug("_get_discoveries")
```
The method is called `_async_get_discoveries`. Personally I'd remove the logs that don't log any data.

MartinHjelmare commented on homeassistant/components/dlna_dms/config_flow.py:
```diff
@@ -0,0 +1,177 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Allow the user to confirm adding the device."""
+        LOGGER.debug("async_step_confirm: %s", user_input)
```
The user input here is not relevant as it's just a confirm step. I'd remove this log.

MartinHjelmare commented on tests/components/dlna_dms/test_config_flow.py:
```diff
@@ -0,0 +1,346 @@
+    }
+    assert result["options"] == {}
+
+    await hass.async_block_till_done()
```
Please block till done directly under the last `async_configure` call.

MartinHjelmare commented on tests/components/dlna_dms/test_config_flow.py:
```diff
@@ -0,0 +1,346 @@
+    assert result["errors"] is None
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
```
We normally patch the integration setup function on all create entry results to avoid setting up the whole integration. See the scaffold script test template for example.

MartinHjelmare commented on tests/components/dlna_dms/conftest.py:
```diff
@@ -0,0 +1,131 @@
+    with patch(
+        "homeassistant.components.dlna_dms.dms.AiohttpSessionRequester", autospec=True
+    ):
+        yield get_domain_data(hass)
```
We don't want to interact with integration details in the tests. We should aim at asserting the library mocks instead.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

MartinHjelmare commented on tests/components/dlna_dms/test_media_source.py:
```diff
@@ -0,0 +1,255 @@
+@pytest.fixture
+async def dms_source(hass: HomeAssistant, entity: DmsDeviceSource) -> DmsMediaSource:
+    """Fixture providing a pre-constructed DmsMediaSource with a single device."""
+    return DmsMediaSource(hass)
```
We should not interact with integration details. See the motioneye media browser tests for examples how to test this platform.

MartinHjelmare commented on tests/components/dlna_dms/test_media_source.py:
```diff
@@ -0,0 +1,255 @@
+    domain_data_mock: DlnaDmsData,
+) -> DmsDeviceSource:
+    """Fixture to set up a DmsDeviceSource in a connected state and cleanup at completion."""
+    await hass.config_entries.async_add(config_entry_mock)
```
Please add the entry with `MockConfigEntry.add_to_hass` and then call `hass.config_entries.async_setup` instead.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Many of the tests are interacting with integration details which we don't want to do. Please rewrite them and aim at asserting core state or library mocks.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations
